#181414763 (feature): Admin CRUD

itisWasp opened the pull request:
### What does this PR do?
Enable Travel admin Create, update, get and delete accommodations
### Description of Task to be completed?
- Restrict the deletion of SUPER_ADMIN role in the system
- CRUD on Accomodation facilities and Location with their Rooms.

### How should this be manually tested?
Get started By Running commands bellow:
  - clone repo
  - switch to Ft-admin-crud-181414763 branch
  - ``yarn install``: to install all dependencies
Test if you can access data stored in your database:
  - Create .env file
  - add a database URL using "TEST_DB" as variable
  - ``yarn migrate``: to migrate table
  - ``yarn test``: to run the test suite

### Any background context you want to provide?
### What are the relevant pivotal tracker stories?
### Screenshots (if appropriate)
### Questions:

sndicunguye reviewed (commented):
Some files still needs to be changed to ES6. There is a mixture of ES5 and ES6 in your codes and I think we should only be using ES6

ericus123 commented:
@itisWasp please rebase and finish this taks ASAP

ericus123 commented on src/controllers/accommodationsController.js:
```diff
@@ -0,0 +1,209 @@
+      // getting user ID from access Token
+      // const token = req.headers.authorization.split(' ')[1];
+      // const decoded = await decodeAcessToken(token);
+      // req.body.userId = decoded.id;
```
remove this

ericus123 commented on src/controllers/addAccomodation.js:
```diff
@@ -0,0 +1,131 @@
+    //   message: 'Accommodation created successfully',
+    //   accommodationInfo,
+    // });
+  };
```
remove unused codes

ericus123 commented on src/controllers/addAccomodation.js:
```diff
@@ -0,0 +1,131 @@
+    return res.status(200).json({
+      message: 'Accommodation deleted successfully',
+    });
+  };
```
On delete also delete the assets associated with the accommodations like images

ericus123 commented on src/database/migrations/20220425063140-create-rooms.js:
```diff
@@ -0,0 +1,42 @@
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      roomType: {
+        type: Sequelize.STRING,
+      },
```
This is supposed to be an ENUM type and have pre-defined room types

ericus123 commented on src/routes/accommodationsRoutes.js:
```diff
@@ -0,0 +1,41 @@
+  verifyAuth,
+  checkAdmin,
+  upload.array('images', 5),
+  // verifyAccomodation,
```
why is this commented out???

ericus123 commented on src/services/accommodationServices.js:
```diff
@@ -0,0 +1,60 @@
+    });
+    if (deletedAccommodation) {
+      return 'Accommodation deleted successfully';
+      // eslint-disable-next-line no-else-return
```
don't disable eslint rules

niyojuste commented on src/controllers/addAccomodation.js:
```diff
@@ -0,0 +1,131 @@
+const { Accommodation, Room } = models;
+const { uploadAccommodationPic } = CloudinaryUtils;
+
+class CrudAccommodation {
```
Are you using this class anywhere?
If **yes**, it appears this is the second class that handles accommodation services. If both classes are necessary, try and improve on the naming of this class as well as this filename.
If **no**, then I'm guessing this whole file is redundant and you don't actually need it

niyojuste commented on src/middlewares/adminAuth.js:
```diff
@@ -0,0 +1,18 @@
+      next();
+    } catch (error) {
+      res.status(400).json({ msg: 'Invalid Token' });
+    }
```
This try..catch is not necessary. Your `if` condition is correctly handling your method.
Also, re-enable this eslint rule and use a **strict** compare and capitalize the ClassName.

niyojuste commented on src/services/accommodationServices.js:
```diff
@@ -0,0 +1,60 @@
+      // eslint-disable-next-line no-else-return
+    } else {
+      return 'Accommodation does not exists';
+    }
```
Where you're calling this `deleteAccommodation` method, first call the `getOneAccommodation` method to handle this error

niyojuste commented on src/swagger/accomodation.doc.js:
```diff
@@ -0,0 +1,192 @@
+      security: [],
+      summary: 'This will retrieve accomodations in given location',
+      description: ' ',
+      produces: ['application/json'],
```
The `produces` field is an OpenApi 2 field. For OpenApi 3 we use the `responses` field only. Find other instances where you used this and remove them

niyojuste commented on src/swagger/accomodation.doc.js:
```diff
@@ -0,0 +1,192 @@
+export const accomodations = {
+  '/api/accommodations/': {
+    post: {
+      tags: ['Accommodations CRUD Operations'],
```
May I suggest naming this tag `Accommodation (Admin only)` or just `Acccommodation (Admin)` as an alternative.

niyojuste commented on src/helpers/token.js:
```diff
@@ -1,8 +1,19 @@
 import jwt from 'jsonwebtoken';
-import { config } from 'dotenv';
+import dotenv from 'dotenv';
 import redis from '../database/redis';
 
-config();
+dotenv.config();
+
+const ACCESS_TOKEN_KEY = process.env.TOKEN_SECRET;
+
+export const decodeAcessToken = async (token) => {
+  try {
+    const decoded = await jwt.verify(token, ACCESS_TOKEN_KEY);
+    return decoded;
+  } catch (err) {
+    console.log(err);
+  }
+};
```
This function you created is basically repeating `verifyToken` below so it's not needed (you also needed to throw the error caught). Also, you can actually remove the dotenv/config import altogether since it's entry point is in the connection to the database at the start of the application (in _src/database/config/config.js_)

niyojuste committed:
ft(admin-crud): Accommodation supplier operations

- Create models (accomodation,facility, facilitycomponent) for implementing the db schema
- Create services and controllers to interact with the models
- Create a middleware for validating the request against invalid data through joi validations logic
- Swagger documentation for each api end point created
- Administrator user should login to be permitted for all operations

[Starts #181414757]

Robertishimwe commented on src/controllers/roomController.js:
```diff
@@ -0,0 +1,97 @@
+      await roomService.updateRoom(req.params.roomId, req.body);
+
+      // if (updatedRoom[0] === 0) {
+      //   return res.status(404).json({
```
why did you comment this out?
if you no longer need it remove it

Robertishimwe replied:
if you are deleting it, then i think you can work on test coz i can see tests are failing then you PRget merged.

niyojuste replied:
Those lines of codes were returning an error upon not finding any room while an empty array is indeed a valid response.

Robertishimwe reviewed (commented):
are you deleting image from cloudinary after deleting accommodation?
